third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs

```diff
@@ -0,0 +1,348 @@
+        &mut self,
+        dst: TempIndex,
+        src: TempIndex,
+        borrowed_locals: &BTreeSet<TempIndex>,
+    ) {
+        if src == dst {
+            // No need to make a copy available for self-assignments.
+            return;
+        }
+        // Note that we are conservative here for the sake of simplicity, and disallow copies
+        // when either `dst` or `src` is borrowed. We could track more copies as available by using
+        // reference analysis.
+        if !borrowed_locals.contains(&dst) && !borrowed_locals.contains(&src) {
+            let old_src = self.0.insert(dst, src);
+            if let Some(old_src) = old_src {
+                panic!(
+                    "ICE: copy `$t{} = $t{}` already available, \
+                        cannot have `$t{} = $t{}` available as well",
+                    dst, old_src, dst, src
+                );
+            }
+        }
+    }
+
+    /// Kill all available copies of the form `x := y` where `x` or `y` is `tmp`.
+    /// If `tmp` is in `borrowed_locals`, then no copies are killed.
+    fn kill_copies_with(&mut self, tmp: TempIndex, borrowed_locals: &BTreeSet<TempIndex>) {
+        if !borrowed_locals.contains(&tmp) {
+            // TODO: consider optimizing the following operation by keeping a two-way map between
+            // `dst -> src` and `src -> set(dst)`. Another optimization to consider is to use im::OrdMap.
+            self.0.retain(|dst, src| *dst != tmp && *src != tmp);
+        }
+    }
+
+    /// Given a set of available copies: `tmp_1 := tmp_0, tmp_2 := tmp_1,..., tmp_n := tmp_n-1`, forming
+    /// the copy chain: `tmp_0 --copied-to--> tmp_1 --copied-to--> tmp_2 -> ... -> tmp_n-1 -> tmp_n`,
+    /// return the head of the copy chain `tmp_0` for any input `tmp_x` (x in 0..=n) in the chain.
+    ///
+    /// Note that it is a required invariant that the copy chain is acyclic, else we panic.
+    /// The natural way of constructing the copy chain for move bytecode (like in this file) ensures this.
+    pub fn get_head_of_copy_chain(&self, mut tmp: TempIndex) -> TempIndex {
+        let mut visited = BTreeSet::from([tmp]);
+        while let Some(src) = self.0.get(&tmp) {
+            if !visited.insert(*src) {
+                // The copy chain is cyclic, which is an invariant violation.
+                panic!("ICE: copy chain is cyclic");
+            }
+            tmp = *src;
+        }
+        tmp
+    }
+}
+
+impl Default for AvailCopies {
+    /// Create a default (empty) collection of definitely available copies.
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AbstractDomain for AvailCopies {
+    /// Keep only those copies in `self` that are available in both `self` and `other`.
+    /// Report if `self` has changed.
+    fn join(&mut self, other: &Self) -> JoinResult {
+        let mut result = JoinResult::Unchanged;
+        let prev_copies = std::mem::take(&mut self.0);
+        for (dst_, src_) in &other.0 {
```

Comment:
Rust has some pretty cool stuff for modifying maps efficiently. Not important here but in general you do:

```
match map.entry(*key) {
  Vacant(handle) => <<insert based on handle>>,
  Occupied(handle) => <<modify or delete based on handle>>
}
```

This way you could instead mutate `self.0` directly instead of reconstructing it.